crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -2,797 +2,535 @@
+    #[tokio::test(flavor = "current_thread", start_paused = true)]
+    async fn test_follower() {
+        telemetry_subscribers::init_for_testing();
 
-    fn random_authority_name() -> AuthorityName {
-        let key: (_, AuthorityKeyPair) = get_key_pair();
-        key.1.public().into()
-    }
+        let (sender, keypair) = get_account_key_pair();
+        let (receiver, _) = get_account_key_pair();
 
-    #[test]
-    fn test_effects_stake() {
-        let mut map = EffectsStakeMap::new();
+        let objects: Vec<_> = (0..10)
+            .map(|_| Object::with_id_owner_for_testing(ObjectID::random(), sender))
+            .collect();
 
-        let threshold = 3;
+        let transactions: Vec<_> = objects
+            .iter()
+            .map(|obj| {
+                make_transfer_sui_transaction(
+                    obj.compute_object_reference(),
+                    receiver,
+                    Some(100),
+                    sender,
+                    &keypair,
+                )
+            })
+            .collect();
+
+        // Set up an authority
+        let config = test_and_configure_authority_configs(1);
+        let authorities = spawn_test_authorities(objects, &config).await;
+        let net = Arc::new(test_authority_aggregator(&config));
+
+        execute_transactions(&net, &transactions).await;
+
+        {
+            debug!("testing follow_one_peer");
+            let sync_store = new_sync_store();
+            let test_handler = TestNodeSyncHandler::new();
+
+            let peer = authorities[0].state().name;
+            let metrics = GossipMetrics::new_for_tests();
+            follow_one_peer(
+                test_handler.clone().break_after(1),
+                sync_store.clone(),
+                peer,
+                net.clone_client(&peer),
+                3,
+                &metrics,
+            )
+            .await
+            .unwrap_err();
+
+            test_handler.clear_break_after();
+
+            let result = follow_one_peer(
+                test_handler.clone(),
+                sync_store.clone(),
+                peer,
+                net.clone_client(&peer),
+                3,
+                &metrics,
+            )
+            .await
+            .unwrap();
 
-        let byzantine = random_authority_name();
-        let validator2 = random_authority_name();
-        let validator3 = random_authority_name();
+            // One item was processed from the db because the first one failed.
+            assert_eq!(result.items_from_db, 1);
 
-        let digests = TransactionEffectsDigest::random();
+            // At least 3 more were fetched from the stream - this number can't be checked exactly
+            // because the authority returns everything in the current batch even if fewer items
+            // were requested, and batch boundaries are non-deterministic.
+            assert!(result.items_from_stream >= 3 && result.items_from_stream <= 5);
```

Comment:
the batch size is 5? any reasons to keep it this small?

Reply:
No, the batch size can be very large, but in low throughput situations new batches are created due to timers, not to exceeding a threshold. So in this case the size of the batch is determined by when the timer fires.

Reply:
but how do we know for sure it's <= 5?

Reply:
empirically, this is what made the test pass. It can never be less than 3, because the test handler won't emit a failure until it sees at least 3. The actual number can be a bit larger than 3, but the important thing is that we fail if it is 10, because that might indicate that the failure was not injected.

I recognize this is not the prettiest test code in the world, but this stuff is very difficult to test and I already spent most of a full day on these tests just to get them to the point they are at.